tests/test_fb_device.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# Copyright (c) 2020 Richard Hull and contributors
+# See LICENSE.rst for details.
+
+"""
+Tests for the :py:class:`luma.core.device.framebuffer` class.
+"""
+import os
+import pytest
+
+from luma.core.render import canvas
+from luma.core.device import fb
+import luma.core.error
+
+from helpers import multi_mock_open
+from unittest.mock import patch, call
+
+SCREEN_RES = "124,55"
+BITS_PER_PIXEL = "24"
+
+
+def test_display_id_as_dev_fb_number():
+    with patch("builtins.open", multi_mock_open(SCREEN_RES, BITS_PER_PIXEL)):
+        device = fb("/dev/fb9")
+        assert device.id == 9
+
+
+def test_display_id_from_environ():
+    os.environ["FRAMEBUFFER"] = "/dev/fb16"
+    with patch("builtins.open", multi_mock_open(SCREEN_RES, BITS_PER_PIXEL)):
+        device = fb()
+        assert device.id == 16
+
+
+def test_unknown_display_id():
+    with patch("builtins.open", multi_mock_open(SCREEN_RES, BITS_PER_PIXEL)):
+        with pytest.raises(luma.core.error.DeviceNotFoundError):
+            fb("invalid fb")
+
+
+def test_read_screen_resolution():
+    with patch(
+        "builtins.open", multi_mock_open(SCREEN_RES, BITS_PER_PIXEL)
+    ) as fake_open:
+        device = fb("/dev/fb1")
+        assert device.width == 124
+        assert device.height == 55
+        fake_open.assert_has_calls([call("/sys/class/graphics/fb1/virtual_size", "r")])
+
+
+def test_read_bits_per_pixel():
+    with patch(
+        "builtins.open", multi_mock_open(SCREEN_RES, BITS_PER_PIXEL)
+    ) as fake_open:
+        device = fb("/dev/fb1")
+        assert device.bits_per_pixel == 24
+        fake_open.assert_has_calls(
+            [call("/sys/class/graphics/fb1/bits_per_pixel", "r")]
+        )
+
+
+def test_display_16bpp():
+    with open("tests/reference/fb_16bpp.raw", "rb") as fp:
```

Comment:
use helper reference file function for this

Reply:
done